[clang] Introduce copy-on-write `CompilerInvocation`

jansvoboda11 opened the pull request:
This PR introduces new copy-on-write `CompilerInvocation` class (`CowCompilerInvocation`), which will be used by the dependency scanner to reduce the number of copies performed when generating command lines for discovered modules.

benlangmuir commented:
I haven't dug into this in depth yet, but did you consider keeping the normal function names for the `const` versions and only rename the mutating ones? My thinking is you could use `CowCompilerInvocation` as a drop-in replacement for `CompilerInvocation` in immutable settings.

jansvoboda11 commented:
I did, but decided against it. There are places in Clang where we store the result of the const function in some kind of long-lived data structure, and someplace else modify the result of the non-const function, expecting the change to be reflected.

I wanted to make the semantic difference clear on both ends in case someone was inclined to swap `CompilerInvocation` for `CowCompilerInvocation`. I'm not sure this is the best way to prevent misuse of the API, so I'm open to alternatives.

Note that you can use `CompilerInvocationBase` to get the behavior you're describing. I use it to implement `generateCC1CommandLine()`, but it's supposed to be used sparingly. Maybe putting it into some kind of `detail` namespace would communicate this even better, but that pollutes the `friend` declarations.

benlangmuir commented:
> There are places in Clang where we store the result of the const function in some kind of long-lived data structure, and someplace else modify the result of the non-const function, expecting the change to be reflected.

I would be comfortable with this, because you would still have to call a distinctly named function (getMut...) in order to do the mutation itself IIUC.  If they are const-casting of course that wouldn't be enough.

> Note that you can use CompilerInvocationBase to get the behavior you're describing.

I think `CompilerInvocationBase` goes a step farther than what I described, because it would give you the mutating methods with the original names, no?

jansvoboda11 commented:
> > There are places in Clang where we store the result of the const function in some kind of long-lived data structure, and someplace else modify the result of the non-const function, expecting the change to be reflected.
> 
> I would be comfortable with this, because you would still have to call a distinctly named function (getMut...) in order to do the mutation itself IIUC. If they are const-casting of course that wouldn't be enough.

Ok, if you're fine with that, I'll use the original names for the const getters on `CowCompilerInvocation`.

> > Note that you can use CompilerInvocationBase to get the behavior you're describing.
> 
> I think `CompilerInvocationBase` goes a step farther than what I described, because it would give you the mutating methods with the original names, no?

No, `CompilerInvocationBase` doesn't have the mutating functions. `RefBase` and `ValBase` only expose the const getters. `CompilerInvocationBase` only adds the `generate*()` functions on top of those. The non-const getters are specific to `CompilerInvocation` and `CowCompilerInvocation`.

jansvoboda11 commented:
I evicted a couple of commits from this PR in order to work better with the new GitHub PR "squash and merge" workflow. I'll create new PRs for the remaining commits.

shafik commented on clang/lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -4588,28 +4641,29 @@ std::string CompilerInvocation::getModuleHash() const {
+  GenerateFileSystemArgs(getFileSystemOpts(), Consumer);
+  GenerateMigratorArgs(getMigratorOpts(), Consumer);
+  GenerateAnalyzerArgs(getAnalyzerOpts(), Consumer);
+  GenerateDiagnosticArgs(getDiagnosticOpts(), Consumer, false);
```
```suggestion
  GenerateDiagnosticArgs(getDiagnosticOpts(), Consumer, /*DefaultDiagColor=*/false);
```

jansvoboda11 replied:
Done in #65647.

shafik reviewed (commented):
nit

benlangmuir commented on clang/include/clang/Frontend/CompilerInvocation.h:
```diff
@@ -66,16 +66,18 @@ bool ParseDiagnosticArgs(DiagnosticOptions &Opts, llvm::opt::ArgList &Args,
-public:
+/// This is a separate base class so that the client-facing sub-classes can
+/// simply declare the desired copy semantics without managing the internals
+/// themselves (for the most parts).
```
Nit: "for the most part" is always with singular "part".

benlangmuir commented on clang/include/clang/Frontend/CompilerInvocation.h:
```diff
@@ -94,47 +96,37 @@ class CompilerInvocationRefBase {
-
-  AnalyzerOptions &getAnalyzerOpts() { return *AnalyzerOpts; }
+public:
+  // clang-format off
```
There seems to be several formatting-off regions here; is this to cheat the length limit? I've never seen formatting disabled like this unless there's a really strong reason.

jansvoboda11 replied:
Yes, I find this hard to read when some of these simple getters fit on one line while others are spread out over three lines.

benlangmuir commented on clang/include/clang/Frontend/CompilerInvocation.h:
```diff
@@ -94,47 +96,37 @@ class CompilerInvocationRefBase {
 
 /// The base class of CompilerInvocation with value semantics.
-class CompilerInvocationValueBase {
+class ValBase {
```
Doesn't have to be this PR, but might be worth just moving everything into RefBase... some of these contain a lot of strings like FrontendOptions and it's not clear ValBase is providing value.

jansvoboda11 replied:
Good point, it would simplify the implementation a bit, too. I'll look into it as a prep patch.

benlangmuir commented on clang/lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -123,49 +123,101 @@ static Expected<std::optional<uint32_t>> parseToleranceOption(StringRef Arg) {
-  AnalyzerOpts.swap(X.AnalyzerOpts);
+namespace clang::CompilerInvocationDetail {
+namespace {
+template <class T> std::shared_ptr<T> make_shared(const T &X) {
```
Why do we have these wrappers?

jansvoboda11 replied:
Because `std::make_shared(const T&)` doesn't deduce the return type as `std::shared_ptr<T>`, so you need to specify `T` explicitly. I find this super verbose, so this utility is here to support invoking the copy constructor of `T` without repeating the type.

benlangmuir replied:
These seem more confusing than helpful; I would either remove them or rename them `make_shared_copy` or similar.

benlangmuir commented on clang/lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -123,49 +123,101 @@ static Expected<std::optional<uint32_t>> parseToleranceOption(StringRef Arg) {
+      PreprocessorOpts(std::make_shared<PreprocessorOptions>()),
+      AnalyzerOpts(llvm::makeIntrusiveRefCnt<AnalyzerOptions>()) {}
+
+RefBase::RefBase(const RefBase &X, DeepCopy)
```
Would it make sense to implement these constructors using `assign`?

jansvoboda11 replied:
I think it would, it'd reduce the boilerplate, and the extra default-construction of the smart pointers is a non-issue.

jansvoboda11 commented on clang/include/clang/Frontend/CompilerInvocation.h:
```diff
@@ -154,46 +146,158 @@ class CompilerInvocationValueBase {
+  using ValBase::getFrontendOpts;
+  using ValBase::getPreprocessorOutputOpts;
+
+  /// RefBase innards
```
This is supposed to span the next 5 declarations. I should probably switch to using Doxygen's grouping mechanism for this (`@{` and `@}`).

jansvoboda11 commented:
Made the entirety of `ValBase` ref-counted in https://github.com/llvm/llvm-project/pull/65647, which also resolves some other comments here.

jansvoboda11 committed:
[clang] Introduce copy-on-write `CompilerInvocation`

When adopted in the dependency scanner, this will speed up execution and reduce memory usage.

benlangmuir reviewed (commented):
This basically LGTM, but could we add some unit tests for the new type?
